users/models.py

```diff
@@ -1,3 +1,246 @@
+        self.role &= ~role_dict[role_name.upper()]
+        if self.role == 0:
+            self.role = self.USER
+        self.save()
+
+    def has_role(self, role_name):
+        """
+        Check if the user has a specific role.
+
+        Args:
+            role_name (str): Name of the role to check.
+
+        Returns:
+            bool: True if the user has the role, False otherwise.
+        """
+        role_dict = {r[1].upper(): r[0] for r in self.ROLE_CHOICES}
+        return self.role & role_dict.get(role_name.upper(), 0) != 0
+
+    def is_admin(self):
+        """Check if the user has admin role."""
+        return bool(self.role & self.ADMIN)
+
+    def has_multiple_roles(self):
+        """Check if the user has multiple roles."""
+        return bin(self.role).count('1') > 1
+
+    def soft_delete(self):
+        """
+        Soft delete the user account.
+
+        This method deactivates the account, sets the deleted_at timestamp,
+        and anonymizes user data.
+
+        Raises:
+            ValidationError: If the account is already inactive or deleted.
+        """
+        if not self.is_active:
+            raise ValidationError("This account is already inactive.")
+        if self.deleted_at:
+            raise ValidationError("This account has already been deleted.")
+        self.is_active = False
+        self.deleted_at = timezone.now()
+        self.original_email = self.email
+        self.email = f"deleted_{self.id}_{self.email}"
+        self.first_name = "Deleted"
+        self.last_name = "User"
+        self.user_phone = ""
+        self.about_me = ""
+        self.profile_picture.delete(save=False)
+        self.save()
+
+    def reactivate(self):
+        """
+        Reactivate a soft-deleted user account.
+
+        This method reactivates the account, clears the deleted_at timestamp,
+        and restores the original email.
+
+        Raises:
+            ValidationError: If the account is already active or not deleted.
+        """
+        if self.is_active:
+            raise ValidationError("This account is already active.")
+        if not self.deleted_at:
+            raise ValidationError("This account was not deleted.")
+        self.is_active = True
+        self.deleted_at = None
+        self.email = self.original_email
+        self.original_email = None
+        self.save()
+
+    def save(self, *args, **kwargs):
+        """
+        Save the user instance to the database.
+
+        This method overrides the default save method to validate the password
+        before saving a new user instance.
+
+        Args:
+            *args: Variable length argument list.
+            **kwargs: Arbitrary keyword arguments.
+        """
+        if self._state.adding and self.password:
+            validate_password(self.password, self)
+        super().save(*args, **kwargs)
+
+    class Meta:
+        verbose_name = 'User'
+        verbose_name_plural = 'Users'
```

Comment:
Since the model uses a custom user system, you might also want to add a custom manager (`CustomUserManager`) for ease of extensibility.